### PR TITLE
Frontend-agnostic 1D scattering

### LINKED ISSUE #870: `scattering(self, x)` could (should?) be frontend-agnostic
Right now, our methods `scattering(self, x)` are frontend-specific, not frontend-agnostic: one is defined in each frontend. Of course, the code in these frontends look very similar: they do argument check, input check, input reshape, scattering, and output reshape.

We have a `self.backend` object, so there is no reason why we couldn't perform all these operations in `base_frontend.py`, with appropriate `shape`, `reshape_input`, and `reshape_output` routines being methods of `self.backend`.

This would help us maintain the code because there would be a single point of entry for the `if out_type="array"` clause and its friends. And i believe it is compliant with #735 philosophy, since it decouples shape pre/post-processing from core scattering.

I have written a short PR where i do this in 1D, to showcase the idea. I surmise that the same would be possible in 2D and 3D.

### PR DESCRIPTION
**Pull request by lostanlen**
segue of #871
if merged, will close #870

negative LOC count
this is an important step towards an automated `self.meta()` algorithm

to manage multiple inheritance, i had to move the `raise NotImplementedError` in kymatio's top-level class to 2D and 3D base frontends. Otherwise the backend method takes precedence because it's first in the inheritance pair. I hope that's OK. Feel free to suggest alternatives

### CODE REVIEW

File: kymatio/scattering1d/frontend/base_frontend.py [hunk 1]
```diff
@@ -82,6 +83,29 @@ def create_filters(self):
             self.J_pad, self.J, self.Q, self.T,
             r_psi=self.r_psi, sigma0=self.sigma0, alpha=self.alpha)
 
+    def scattering(self, x):
+        ScatteringBase1D._check_runtime_args(self)
+        ScatteringBase1D._check_input(self, x)
+        x_shape = self.backend.shape(x)
```
Comment (MuawizChaudhary) [inline kymatio/scattering1d/frontend/base_frontend.py:89]: Add space above this line

**Review by MuawizChaudhary — APPROVED** [review]
I like this. I'll let @janden merge.

**Comment by lostanlen** [comment]
@janden removal of the ~~duplicated~~ (triplicated!) source you found in frontend ;)

File: kymatio/scattering1d/frontend/base_frontend.py (hunk 1 above)
Comment (lostanlen) [reply kymatio/scattering1d/frontend/base_frontend.py:90]: voilà ! 4978b07

**Comment by MuawizChaudhary** [comment]
I am having some second doubts on having this be merged so close to 0.3 release. @lostanlen feel free to speak up if this is something you feel strongly about. I suggest we turn this into a draft for now. What do you think @janden ?

File: kymatio/scattering1d/frontend/base_frontend.py [hunk 2]
```diff
@@ -82,6 +83,32 @@ def create_filters(self):
             self.J_pad, self.J, self.Q, self.T,
             r_psi=self.r_psi, sigma0=self.sigma0, alpha=self.alpha)
 
+    def scattering(self, x):
```
Comment (cyrusvahidi) [inline kymatio/scattering1d/frontend/base_frontend.py:86]: trés cool

File: kymatio/scattering1d/frontend/torch_frontend.py [hunk 3]
```diff
@@ -67,41 +67,8 @@ def load_filters(self):
...
-                S[k] = v.reshape(new_shape)
-        elif self.out_type == 'list':
-            for x in S:
-                x.pop('n')
-                scattering_shape = x['coef'].shape[-1:]
-                new_shape = batch_shape + scattering_shape
-                x['coef'] = x['coef'].reshape(new_shape)
-
-        return S
+        return super().scattering(x)
```
Comment (cyrusvahidi) [inline kymatio/scattering1d/frontend/torch_frontend.py:71]: bon

File: kymatio/scattering1d/frontend/base_frontend.py (hunk 2 above)
Comment (lostanlen) [reply kymatio/scattering1d/frontend/base_frontend.py:86]: wait till this is used to implement `.meta()` and `.out_size()` automatically (thanks to `DryBackend()`) ☺️

File: kymatio/scattering2d/frontend/base_frontend.py [hunk 4]
```diff
@@ -35,6 +35,10 @@ def create_filters(self):
         filters = filter_bank(self.M_padded, self.N_padded, self.J, self.L)
         self.phi, self.psi = filters['phi'], filters['psi']
 
+    def scattering(self, x):
```
Comment (cyrusvahidi) [inline kymatio/scattering2d/frontend/base_frontend.py:38]: is this to prepare 2D for what you have done with 1D?

Comment (cyrusvahidi) [inline kymatio/scattering2d/frontend/base_frontend.py:38]: will this eventually follow the same pattern as `kymatio/scattering1d/frontend/base_frontend.py`?

Comment (lostanlen) [reply kymatio/scattering2d/frontend/base_frontend.py:38]: maybe. if needed!

**Review by janden — CHANGES_REQUESTED** [review]
There seems to be some rebase issues here (the new backend functions from #871 are showing up in the diff). Overall, I think this looks good.

@MuawizChaudhary Do you have any specific concerns here that merit postponing merge until 0.3 is out?

**Comment by lostanlen** [comment]
@janden this PR is built on top of (yet unmerged) #871. that's why the commits of #871 are showing up. #871 should be merged first and this one will appear to reduce in size

**Comment by MuawizChaudhary** [comment]
> There seems to be some rebase issues here (the new backend functions from #871 are showing up in the diff). Overall, I think this looks good.
> 
> @MuawizChaudhary Do you have any specific concerns here that merit postponing merge until 0.3 is out?

I think for me its that the 1d scattering will be mildly different in design from the 2d scattering. I also wonder if the Jax backend will play nicely with these changes.

I'm most definitely not in support of merging these if it leads to delays.

**Comment by cyrusvahidi** [comment]
tests are failing

**Review by cyrusvahidi — CHANGES_REQUESTED** [review]
tests need fixing

**Comment by lostanlen** [comment]
This works in JAX. Tests are passing again.

**Comment by lostanlen** [comment]
@rastegah please review, merge to dev, and then to main like you did in #906

File: kymatio/scattering1d/frontend/numpy_frontend.py [hunk 5]
```diff
@@ -18,36 +18,24 @@ def __init__(self, J, shape, Q=1, T=None, max_order=2, average=True,
     def scattering(self, x):
         ScatteringBase1D._check_runtime_args(self)
         ScatteringBase1D._check_input(self, x)
-
-        batch_shape = x.shape[:-1]
-        signal_shape = x.shape[-1:]
-
-        x = x.reshape((-1, 1) + signal_shape)
+        x_shape = self.backend.shape(x)
+        batch_shape, signal_shape = x_shape[:-1], x_shape[-1:]
```
Comment (cyrusvahidi) [inline kymatio/scattering1d/frontend/numpy_frontend.py]: what if i the input shape is `(batch, channels, signal_shape)`? i feel that this is dangerous, particularly given that `_check_input` is weak and even has a comment to say so: ttps://github.com/lostanlen/kymatio/blob/571e1b3d98c1e8e6ad8bb73a8cd81065bdbd99ec/kymatio/scattering1d/frontend/base_frontend.py#L207

maybe we should strengthen `_check_input`

Comment (cyrusvahidi) [reply kymatio/scattering1d/frontend/numpy_frontend.py]: or maybe i'm wrong?

**Review by cyrusvahidi — APPROVED** [review]

File: kymatio/scattering1d/frontend/numpy_frontend.py (hunk 5 above)
Comment (cyrusvahidi) [reply kymatio/scattering1d/frontend/numpy_frontend.py]: although this was already part of the implementation

**Review by cyrusvahidi — APPROVED** [review]

File: kymatio/frontend/numpy_frontend.py [hunk 6]
```diff
@@ -4,10 +4,6 @@ class ScatteringNumPy:
     def __init__(self):
         self.frontend_name = 'numpy'
 
-    def scattering(self, x):
-        """ This function should compute the scattering transform."""
-        raise NotImplementedError
-
```
Comment (MuawizChaudhary) [inline kymatio/frontend/numpy_frontend.py:10]: Why remove?

**Review by MuawizChaudhary — APPROVED** [review]
Looks good to me. Not a fan of this being applied to just 1d and having the 2d and 3d be ignored, but I suppose we will get to that in due time. Merging.

**Comment by lostanlen** [comment]
cherry-picked and pushed to main

**Pull request merged**